static/js/user_groups.ts

```diff
@@ -1,43 +1,69 @@
 import * as blueslip from "./blueslip";
 import {FoldDict} from "./fold_dict";
+import type {UserGroupUpdatEvent} from "./types";
 
-let user_group_name_dict;
-let user_group_by_id_dict;
+type UserGroup = {
+    description: string;
+    id: number;
+    name: string;
+    members: Set<number>;
+};
+
+// The members field is a number array which we convert
+// to a Set in the initialize function.
+type UserGroupRaw = Omit<UserGroup, "members"> & {members: number[]};
+
+let user_group_name_dict: FoldDict<UserGroup>;
+let user_group_by_id_dict: Map<number, UserGroup>;
 
 // We have an init() function so that our automated tests
 // can easily clear data.
-export function init() {
+export function init(): void {
     user_group_name_dict = new FoldDict();
-    user_group_by_id_dict = new Map();
+    user_group_by_id_dict = new Map<number, UserGroup>();
 }
 
 // WE INITIALIZE DATA STRUCTURES HERE!
 init();
 
-export function add(user_group) {
+export function add(user_group_raw: UserGroupRaw): void {
     // Reformat the user group members structure to be a set.
-    user_group.members = new Set(user_group.members);
+    const user_group = {
+        description: user_group_raw.description,
+        id: user_group_raw.id,
+        name: user_group_raw.name,
+        members: new Set(user_group_raw.members),
+    };
+
     user_group_name_dict.set(user_group.name, user_group);
     user_group_by_id_dict.set(user_group.id, user_group);
 }
 
-export function remove(user_group) {
+export function remove(user_group: UserGroup): void {
     user_group_name_dict.delete(user_group.name);
     user_group_by_id_dict.delete(user_group.id);
 }
 
-export function get_user_group_from_id(group_id, suppress_errors) {
+export function get_user_group_from_id(
+    group_id: number,
+    suppress_errors = false,
+): UserGroup | undefined {
     if (!user_group_by_id_dict.has(group_id)) {
-        if (suppress_errors === undefined) {
+        if (!suppress_errors) {
             blueslip.error("Unknown group_id in get_user_group_from_id: " + group_id);
```

Comment:
Can we just make this error fatal, and have the one caller who wants `suppress_errors` catch the exception? The callers who aren’t expecting `undefined` were going to crash anyway when they try to access, e.g., `group.name`.

Then this can return `UserGroup` and we don’t need to add separate error messages at each call site.

Reply:
Made this error fatal.

Reply:
This wfm.  @m-e-l-u-h-a-n can help with any refactoring of that form.